### PR TITLE
[JIMT][level-editor-not-working] - filters out removed songs

### PR DESCRIPTION
**Pull request by thomasoniii**
If a song was selected for a level and subsequently removed, then the level editor would crash.

The issue was it was trying to look up the name of a song given its ID and failing to do so (the song has been removed), so it'd try to index into a `nil` object.

So we just filter out to objects that are known to exist. That fixes the crash and ensures that the deleted selected song isn't in the list of songs.

Please check my work. A more knowledgeable ruby user could probably tidy this up - I have redundant select calls and I feel like we could tidy this up into a single variable somewhere else further up the chain. But this resolves the issue and should work.
## Links

- level: [http://localhost-studio.code.org:3000/s/dance-ai-2023/lessons/1/levels/2](/s/dance-ai-2023/lessons/1/levels/2)
- trello: [https://trello.com/c/bDXrHv8i/174-levelbuilder-edit-not-working-urgent](Levelbuilder edit not working - urgent)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/views/levels/editors/fields/_default_song.html.haml [hunk 1]
```diff
@@ -6,7 +6,7 @@
   It is defined as: If there are selected songs, then use the first one in the list. Otherwise, use the first song in the list
   of all songs.
 - songMap = @level.class.hoc_songs.reduce({}) {|hash, song| hash[song[1]] = song; hash }
-- selectedSongObjects = @level.song_selection ? @level.song_selection.map {|s| songMap[s] } : @level.class.hoc_songs
+- selectedSongObjects = @level.song_selection ? @level.song_selection.map {|s| songMap[s] }.select{|s| s} : @level.class.hoc_songs
```
Comment (bencodeorg) [inline dashboard/app/views/levels/editors/fields/_default_song.html.haml]: Also not an expert, but I think `.compact` might be the more standard to remove nil values in Ruby? [Docs](https://ruby-doc.org/core-1.9.3/Array.html#method-i-compact)

File: dashboard/app/views/levels/editors/fields/_song_selection.html.haml [hunk 2]
```diff
@@ -8,7 +8,7 @@
     Click to
     %a.select_none{href: '#'} not limit song selection
   %p (shift-click or cmd-click to select multiple).
-  = f.select :song_selection, ((@level.song_selection || []).map {|s| songMap[s] } + @level.class.hoc_songs).uniq, {}, {size: 12, multiple: true}
+  = f.select :song_selection, ((@level.song_selection || []).map {|s| songMap[s] }.select {|s| s} + @level.class.hoc_songs).uniq, {}, {size: 12, multiple: true}
   %p Songs currently selected (does not update until level is saved):
 
-  = @level.song_selection ? @level.song_selection.map {|s| songMap[s][0] } : "All songs are available"
+  = @level.song_selection ? @level.song_selection.map {|s| songMap[s]&.[](0) }.select {|s| s} : "All songs are available"
```
Comment (bencodeorg) [inline dashboard/app/views/levels/editors/fields/_song_selection.html.haml]: I know this is haml, which doesn't lend itself to readability, but maybe something a little more verbose like `songMap[s]&.length && songMap[s][0]` might be a little easier to understand? I see the string `&.[` only a few times in our repo (and never to grab the first element of an array)

**Review by bencodeorg — APPROVED** [review]
Couple style ideas but LGTM!

File: dashboard/app/views/levels/editors/fields/_default_song.html.haml (hunk 1 above)
Comment (thomasoniii) [reply dashboard/app/views/levels/editors/fields/_default_song.html.haml]: That's the kinda ruby knowledge I lack! tyvm - changed it.

File: dashboard/app/views/levels/editors/fields/_song_selection.html.haml (hunk 2 above)
Comment (thomasoniii) [reply dashboard/app/views/levels/editors/fields/_song_selection.html.haml]: And this one I'll defer pre-existing styles, so changed as well.

I definitely agree that It's a pretty weird looking syntax.

**Review by bencodeorg — APPROVED** [review]
Looks good to me! Just to confirm, did you take a quick manual pass to make sure things look good with my suggestions? My Ruby is rusty, just want to make sure I didn't lead you astray :)

**Comment by bencodeorg** [comment]
~~Oh I definitely led you astray :) I forgot !!0 is true in Ruby. You can use `empty?` though, [docs](https://apidock.com/ruby/Array/empty%3F)~~ nvm, length will work (although the !!0 thing is true), I'm all over the place. Apologies!

**Review by bencodeorg — APPROVED** [review]
LGTM!

**Review by fisher-alice — COMMENTED** [review]
This looks great! Thanks for taking care of this so quickly.

**Pull request merged**